### PR TITLE
Unit test for GATKReadCoder

### PR DESCRIPTION
**Pull request by droazen**
Added a unit test to verify that GATKReads are being coded correctly
regardless of backing implementation.

### CODE REVIEW

**Comment by droazen** [comment]
For @davidadamsphd to review

File: src/test/java/org/broadinstitute/hellbender/engine/dataflow/coders/GATKReadCoderUnitTest.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+import java.util.UUID;
+
+
+public class GATKReadCoderUnitTest extends BaseTest {
+
+    private GATKRead makeGoogleRead( final int uuid, final String name, final int start, final int length ) {
+        return ArtificialReadUtils.createGoogleBackedReadWithUUID(new UUID(0, uuid), name, start, length);
+    }
+
+    public GATKRead makeSamRead( final int uuid, final String name, final int start, final int length ) {
```
Comment (davidadamsphd) [inline src/test/java/org/broadinstitute/hellbender/engine/dataflow/coders/GATKReadCoderUnitTest.java]: public?
Also, FYI, these will get more to a common util after the read pre-processing pipeline is ready.

File: src/test/java/org/broadinstitute/hellbender/engine/dataflow/coders/GATKReadCoderUnitTest.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    }
+
+    @DataProvider(name = "reads")
+    public Object[][] makeReads() {
+        final List<GATKRead> googleReads = Arrays.asList(makeGoogleRead(1, "google1", 1, 10), makeGoogleRead(2, "google2", 20, 40));
+        final List<GATKRead> samReads = Arrays.asList(makeSamRead(1, "sam1", 1, 10), makeSamRead(2, "sam2", 20, 40));
+
+        return new Object[][] {
+                { googleReads },
+                { samReads }
```
Comment (davidadamsphd) [inline src/test/java/org/broadinstitute/hellbender/engine/dataflow/coders/GATKReadCoderUnitTest.java]: Create a mixed test?

**Comment by davidadamsphd** [comment]
LGTM, please add the addition test if you agree, otherwise merge at will.

File: src/test/java/org/broadinstitute/hellbender/engine/dataflow/coders/GATKReadCoderUnitTest.java (hunk 1 above)
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/engine/dataflow/coders/GATKReadCoderUnitTest.java]: Was a typo -- fixed.

File: src/test/java/org/broadinstitute/hellbender/engine/dataflow/coders/GATKReadCoderUnitTest.java (hunk 2 above)
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/engine/dataflow/coders/GATKReadCoderUnitTest.java]: Added one, but now I get an error: `com.google.cloud.dataflow.sdk.coders.CannotProvideCoderException: Cannot provide coder for Create: The elements are not all of the same class.`

Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/engine/dataflow/coders/GATKReadCoderUnitTest.java]: After further investigation, I don't think the approach here to testing the `GATKReadCoder` is legitimate -- the (non-mixed) tests pass even if I comment out the line `p.getCoderRegistry().registerCoder(GATKRead.class, new GATKReadCoder());` in `DataFlowUtils.registerGATKCoders()` (which causes catastrophic errors in most of our other dataflow tests).

Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/engine/dataflow/coders/GATKReadCoderUnitTest.java]: If I set the coder explicitly for the `Create` operation it seems to work (even in the mixed implementation case!). Need to discuss with you the implications of this.

**Comment by droazen** [comment]
@davidadamsphd I've pushed a new version that properly tests the case of a `PCollection` with a mix of different `GATKRead` implementations. I've also verified in a debugger that the new version is actually invoking the `GATKReadCoder` properly. Since there are some interesting questions raised about the need to call `setCoder()` manually on our transforms, I'd like to request a re-review to get your thoughts.

File: src/test/java/org/broadinstitute/hellbender/engine/dataflow/coders/GATKReadCoderUnitTest.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+        // of that type and see if it matches the List version.
+        final Pipeline p = TestPipeline.create();
+        DataflowUtils.registerGATKCoders(p);
+
+        // Need to explicitly set the coder to GATKReadCoder, otherwise Create fails to infer
+        // a coder properly in the case where the List contains a mix of different GATKRead implementations.
+        final PCollection<GATKRead> dataflowReads = p.apply(Create.of(reads)).setCoder(new GATKReadCoder());
+        DataflowAssert.that(dataflowReads).containsInAnyOrder(reads);
+
+        final PCollection<GATKRead> dataflowReadsAfterTransform = dataflowReads.apply(ParDo.of(new DoFn<GATKRead, GATKRead>() {
```
Comment (davidadamsphd) [inline src/test/java/org/broadinstitute/hellbender/engine/dataflow/coders/GATKReadCoderUnitTest.java]: Please create an issue (and file it against me) to move this usage to a more canonical utility (based on testing cod that's now in the SDK).

**Comment by davidadamsphd** [comment]
LGTM. Please merge after creating the issue and updating the comment. 
As for the setCoder, I think we're going to see that explicitly setting a coder on Create.of is common. We always need it for construction of collections of collections. So, I think what you did is fine.

**Pull request merged**